### PR TITLE
Bug 1786675: OpenStack: create security rules sequentially

### PR DESCRIPTION
**Pull request by Fedosin**
Terraform tries to create all rules at once, sending multiple requests simultaneously. While this can reduce installation time a bit, it also can lead to conflicts. It is especially evident on high loaded systems.

Terraform developers recommend a workaround to create rules sequentially by setting explicit dependencies between them:
https://github.com/hashicorp/terraform/issues/7519
This commit fixes the issue according to the workaround recommendation.

When https://github.com/terraform-providers/terraform-provider-openstack/issues/952 is fixed in the to the OpenStack Terraform provider, we can revert this patch.

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
I would like to see if we can improve the openstack provider upsteeam such that 
- it internally serializes actions on these resources
- add retries on conflicts

then we can carry this fix for time being under the assumptions that at some point we can dorp this hack for an upstream fix..

Also the linked issue is from 2016, maybe we can at least open an issue on https://github.com/terraform-providers/terraform-provider-openstack to track the need ??

**Comment by mandre** [comment]
Let's be clear this is a workaround for a defect in OpenStack terraform provider. Ultimately, this is where the issue should be fixed. Please open an issue at https://github.com/terraform-providers/terraform-provider-openstack as recommended by @abhinavdahiya.
Now, regarding the proposed fix, I understand this is blocking the installation of OCP on slower OpenStack clouds and we need a fix urgently, but did you measure by how much longer it takes to create the SGs in serial vs parallel? This would help decide whether this is an acceptable trade-off and if we should prioritize fixing the issue in the openstack tf provider.

**Comment by mandre** [comment]
I just ran some tests and serializing the SGs creation isn't that bad, 36 sec vs 12 sec on Mass Open Cloud. By default, terraform runs up to 10 operations in parallel.

Parallel SGs creation from master:
```
Apply complete! Resources: 48 added, 0 changed, 0 destroyed.
~/go/bin/terraform apply -auto-approve  3.16s user 0.50s system 30% cpu 11.996 total
```

Serial SGs creation, with the patch applied:
```
Apply complete! Resources: 48 added, 0 changed, 0 destroyed.
~/go/bin/terraform apply -auto-approve  3.21s user 0.61s system 10% cpu 36.341 total
```

**Comment by Fedosin** [comment]
Unfortunately this is a test blocker for us - we can't deploy OCP on PSI because of this issue. So, I think we should merge the workaround and then try to fix it in the the openstack tf provider.

**Comment by mandre** [comment]
> Unfortunately this is a test blocker for us - we can't deploy OCP on PSI because of this issue. So, I think we should merge the workaround and then try to fix it in the the openstack tf provider.

Agreed - the request was to create an issue in https://github.com/terraform-providers/terraform-provider-openstack and reference it from this commit.

**Comment by Fedosin** [comment]
also, we should keep in mind that we need to backport this to 4.3 and 4.2

**Comment by adduarte** [comment]
I do have a question, this patch simply slows down things by making sure that the previous rule has been created. But would it actually fix the problem which is caused by the failure of terraform handling the error?   is there a case where the connection is so bad that even this patch will fail, because terraform tries to create the rule, something messes up the communication between the cloud and terrform, so terraform tries to create it again and runs into same issue? Is there a way to ask terraform to check and see if a rule exists before trying to create it? that could get around this problem.

**Comment by Fedosin** [comment]
It's not exactly about connection problems, this concerns the load on the test cloud.... The point is that now we send 10 simultaneous rule creation requests and Neutron is unable to process all of them within a given amount of time.
This leads to the fact that after a timeout Terraform believes that the creation failed (no response from Neutron) and sends another request, which leads to a conflict.
To tell you the truth, this is not quite a fix, but a workaround, as there is still a possibility of conflict even with a single request. But it significantly reduces the probability of such a development.
In other words, we reduce the load on Neutron to a possible minimum, below which we are unable to install the system anyway. That is, if Neutron can't handle even a single request, then the problem is exactly in the cloud itself, not us.
